### ISSUE TITLE
Question about apk's assets folder and source code in it

### ISSUE THREAD

**Issue by hamidrezabstn** [NativeScript/NativeScript#2946]
hi 

i'm new to nativescript ! after i create grocery sample app ,i've build it and then i open the apk output using a simple archive manager AND BOOM! i see all the source code inside assets folder! i was shocked! 
is that normal ? anything wrong with my build ?

**Comment by NathanaelA** [comment]
Hi,
  
    This is actually a known issue with ALL of the JavaScript frameworks,  ReactNative, Titanium, FuseJS, Ionic/Cordova/Phoegap.   All of them you can extract the JS code out of them.    .     Now, I have some good news for you; There is an already released product (Disclosure: I'm the developer) call https://AppProtection.net that will encrypt the source.

**Comment by hamidrezabstn** [comment]
@NathanaelA thanks for your kind response
is there any open source solution ? :smile:

**Comment by bradmartin** [comment]
Doesn't this happen with native Android apps also, unless you use proguard or some encryption utility? I think I've opened some native apps and seen the java source.

**Comment by hamidrezabstn** [comment]
@bradmartin exactly its not that easy ! 

in native apk you have compiled classes and dex files ! and yes  , with some tools , you can extract source code which with help of proguard , they are hard to read and use .

 but extracting from nativescript apk is easy **as copy and paste.**

**Comment by NathanaelA** [comment]
@hamidrezabstn - unless you use a commercial tool on the Java side it is trivial to extract the sources and pull any valuable data out, the source **might** be a little harder to read with proguard; but most the time not very much more, i personally have no issues with proguard'd apps....  ;-)      DexGuard (commercial) does a much better job, but again it isn't perfect.

Now as to NativeScript encryption; Telerik is coming out with a open source plugin to encrypt the app soon ( not sure the release date yet ).     I have already been able to evaluate it and it offers about as much protection as the cordova/phonegap/ionic encryption plugin does; which is to say if you hand me your app I can decrypt all the source code in under a minute.  ;-)    It will protect against a script kiddie; but any decent developer that has a couple hours to waste can figure out a couple ways to decrypt cordova or telerik's encryption system trivially...

So basically Telerik's is like a simple proguard, and mine is like dexguard.  ;-)

**Comment by bradmartin** [comment]
Yea I know about NativeScript and all the other hybrids being in the open.
I've opened them myself plenty and a few native apps. Just not compiled any
native to know that process too well.

**Comment by enchev** [comment]
Hey @hamidrezabstn , 
We can actually enable UglifyJS plugin for Webpack. Will this work for you?

**Comment by hamidrezabstn** [comment]
Hi @enchev 

yes it should help ! could you explain it a bit more ? 
or give me some reference for learning it !

**Comment by NickIliev** [comment]
@hamidrezabstn 
Nice blog article about using webpack and uglify [here](https://medium.com/@enchev/minify-uglify-and-bundle-your-javascript-in-nativescript-angular-2-app-aabf84f47fbb#.57xqaxcid)
